File: Centrality_Analysis/closeness.cpp
#include <bits/stdc++.h>
using namespace std;

int dijkstra(vector<vector<int>>& adj, int src, int dest) {
    int n = adj.size();
    vector<int> dist(n, INT_MAX);
    dist[src] = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, src});

    while(!pq.empty()){
        int d = pq.top().first;
        int node = pq.top().second;
        pq.pop();

        if(node == dest) return d;

        for(int adjNode: adj[node]){
            if(d + 1 < dist[adjNode]){
                dist[adjNode] = d + 1;
                pq.push({dist[adjNode], adjNode});
            }
        }
    }
    return -1;
}

int main(){
    vector<vector<int>> adj={{2,4},{2},{0,1,3},{2,4},{0,3}};
    int n = adj.size();

    double maxi = 0;
    vector<int> node_ans;

    for(int i=0; i<n; i++){
        int dist = 0;
        for(int j=0; j<n; j++){
            if(i==j) continue;
            int dist_ij = dijkstra(adj, i, j);
            dist += dist_ij;
        }
        double closeness_centrality = (n-1) / (double)dist;
        if(closeness_centrality > maxi) {
            maxi = closeness_centrality;
            node_ans.clear();
            node_ans.push_back(i);
        } else if(closeness_centrality == maxi) {
            node_ans.push_back(i);
        }
        cout << "Closeness Centrality of node " << i << ": " << closeness_centrality << endl;
    }

    cout << "Node(s) with maximum closeness centrality: ";
    for(auto node : node_ans) {
        cout <<"Node "<<node<< endl;
    }
}

File: Centrality_Analysis/closeness.py
import heapq

def dijkstra(adj, src, dest):
    n = len(adj)
    dist = [float('inf')] * n
    dist[src] = 0
    pq = [(0, src)]

    while pq:
        d, node = heapq.heappop(pq)

        if node == dest:
            return d

        for adjNode in adj[node]:
            if d + 1 < dist[adjNode]:
                dist[adjNode] = d + 1
                heapq.heappush(pq, (dist[adjNode], adjNode))

    return -1


adj = [[2, 4], [2], [0, 1, 3], [2, 4], [0, 3]]
n = len(adj)

maxi = 0
node_ans = []

for i in range(n):
    dist_sum = 0
    for j in range(n):
        if i == j:
            continue
        dist_ij = dijkstra(adj, i, j)
        dist_sum += dist_ij

    closeness_centrality = (n - 1) / dist_sum

    if closeness_centrality > maxi:
        maxi = closeness_centrality
        node_ans = [i]
    elif closeness_centrality == maxi:
        node_ans.append(i)

    print(f"Closeness Centrality of node {i}: {closeness_centrality}")

print("Node(s) with maximum closeness centrality:")
for node in node_ans:
    print(f"Node {node}")